crates/agent/src/api/authorize_user_prefix.rs

```diff
@@ -0,0 +1,520 @@
+    started_unix: u64,
+) -> Result<axum::Json<Response>, crate::api::ApiError> {
+    let started = chrono::DateTime::from_timestamp(started_unix as i64, 0).unwrap_or_default();
+
+    match Snapshot::evaluate(snapshot, started, |snapshot: &Snapshot| {
+        evaluate_authorization(
+            snapshot,
+            user_id,
+            email.as_ref(),
+            &prefix,
+            &data_plane,
+            capability,
+        )
+    }) {
+        Ok((
+            exp,
+            (encoding_key, mut broker_claims, broker_address, mut reactor_claims, reactor_address),
+        )) => {
+            broker_claims.inner.exp = exp.timestamp() as u64;
+            broker_claims.inner.iat = started.timestamp() as u64;
+            reactor_claims.inner.exp = exp.timestamp() as u64;
+            reactor_claims.inner.iat = started.timestamp() as u64;
+
+            let header = jsonwebtoken::Header::default();
+            let broker_token = jsonwebtoken::encode(&header, &broker_claims, &encoding_key)
+                .context("failed to encode authorized JWT")?;
+            let reactor_token = jsonwebtoken::encode(&header, &reactor_claims, &encoding_key)
+                .context("failed to encode authorized JWT")?;
+
+            Ok(axum::Json(Response {
+                broker_token,
+                broker_address,
+                reactor_token,
+                reactor_address,
+                retry_millis: 0,
+            }))
+        }
+        Err(Ok(backoff)) => Ok(axum::Json(Response {
+            retry_millis: backoff.as_millis() as u64,
+            ..Default::default()
+        })),
+        Err(Err(err)) => Err(err),
+    }
+}
+
+fn evaluate_authorization(
+    snapshot: &Snapshot,
+    user_id: uuid::Uuid,
+    user_email: Option<&String>,
+    prefix: &models::Prefix,
+    data_plane_name: &models::Name,
+    capability: models::Capability,
+) -> Result<
+    (
+        Option<chrono::DateTime<chrono::Utc>>,
+        (
+            jsonwebtoken::EncodingKey,
+            super::DataClaims, // Broker claims.
+            String,            // Broker address.
+            super::DataClaims, // Reactor claims.
+            String,            // Reactor address.
+        ),
+    ),
+    crate::api::ApiError,
+> {
+    if !tables::UserGrant::is_authorized(
+        &snapshot.role_grants,
+        &snapshot.user_grants,
+        user_id,
+        prefix,
+        capability,
+    ) {
+        return Err(anyhow::anyhow!(
+            "{} is not authorized to {prefix} for {capability:?}",
+            user_email.map(String::as_str).unwrap_or("user")
+        )
+        .with_status(StatusCode::FORBIDDEN));
+    }
+
+    if !tables::UserGrant::is_authorized(
+        &snapshot.role_grants,
+        &snapshot.user_grants,
+        user_id,
+        &data_plane_name,
+        models::Capability::Read,
+    ) {
+        return Err(anyhow::anyhow!(
+            "{} is not authorized to {data_plane_name}",
```

Comment:
Not sure if we care about being able to guess private data-plane names, but this message+status code being different from the "not found" error below would let a potential bad actor check whether a guessed data-plane exists or not.

Reply:
How? We're verifying their authorization to the requested data-plane name **before** we check whether the name actually exists. They will always get a FORBIDDEN if they don't have RBAC access, regardless of whether it exists.